### ISSUE TITLE
HtmlEditorExtender InsertImage error in IIS

### ISSUE THREAD

**Issue by rclancy** [DevExpress/AjaxControlToolkit#168]
I am facing identical errors to #43. The control works fine in IIS Express but when hosted locally in IIS an error is produced when clicking Upload. It is the same error shown in several other posts - like 43 above.

I'm using AjaxControlToolkit v16.1 in an asp.net v4.6.1 project. Setting the trust level to Full doesn't help. UseAbsoluteHandlerPath is not available as a control attribute. I can see the value being set in the js debug file, but not in the release version. Adding the IIS app name to the AjaxFileUploadHandler.axd path in web.config also doesn't help.

The console error is: the server cannot find the requested URI: http://localhost/AjaxFileUploadHandler.axd?.. It is interesting that this is the same error even after changing the path in web.config, suggesting that UseAbsoluteHandlerPath is set to true in release. Maybe the problem is that I simply don't know how to set that flag, apart from overwriting it in AjaxFileUpload.debug.js, which causes the build to fail.

**Comment by AlekseyMartynov** [comment]
> UseAbsoluteHandlerPath is not available as a control attribute. 

It is a property of AjaxFileUpload control, find it in the Misc category of the Properties window:

![afu](https://cloud.githubusercontent.com/assets/4426185/16761562/583174b2-4829-11e6-97e2-b66ab26a12ee.png)

[(Source code)](https://github.com/DevExpress/AjaxControlToolkit/blob/16.1.0/AjaxControlToolkit/AjaxFileUpload/AjaxFileUpload.cs#L112-L118)

> The console error is: the server cannot find the requested URI: http://localhost/AjaxFileUploadHandler.axd?.. 

Make sure you have the handler registered in web.config ([example](https://github.com/DevExpress/AjaxControlToolkit/blob/16.1.0/AjaxControlToolkit.SampleSite/Web.config#L26-L30))

**Comment by rclancy** [comment]
How can I access the properties window of the AjaxFileUpload control? I can't seem to find it. Do I need to load the control into the Toolbox or is there some way of selecting it in the Web Form's Design mode (which I have no experience with, working only in Source mode)? Sorry for asking such a basic question. 

The handler is registered in web.config.

**Comment by AlekseyMartynov** [comment]
It's available in Source mode too, no need to use Designer:

![untitled](https://cloud.githubusercontent.com/assets/4426185/16798173/8e13aef4-48f3-11e6-8e02-daab933bee19.png)

If the option is absent, double check the version of AJAX Control Toolkit referenced to the project. My screenshots are made with 16.1.0

**Comment by rclancy** [comment]
I'm sorry, I've made a mistake. I'm having this exact same problem as described above but with ajaxToolkit:HtmlEditorExtender Toolbar:InsertImage, not AjaxFileUpload. As far as I can see, the path attribute is not available for the HtmlEditorExtender. Otherwise the scenario is identical. 

I'm also using asp:ScriptReference which references the AjaxControlToolkit/Bundle and my project is using v16.1.0 of the Toolkit.

**Comment by AlekseyMartynov** [comment]
Would you provide us with minimal steps to reproduce the problem so we could debug it on our side? A minimal sample app in a ZIP archive and the description of your web server configuration (pool mode, root path, other specifics) would be ideal.

Check our [guidelines for submitting bugs](https://github.com/DevExpress/AjaxControlToolkit/blob/master/CONTRIBUTING.MD).
Also please update the ticket title to reflect the actual issue.
Thanks.

**Comment by rclancy** [comment]
I've created a sample project in VS2015, 16.1.0, Win10 Pro:
- no master page is used, whereas my original app uses master pages
- the error occurs in IIS express, before deployment, so I haven't taken it any further until this is diagnosed. With my original app, the control works perfectly in IIS Express but fails when deployed to IIS (localhost)
- browser: IE

[HtmlEditor.zip](https://github.com/DevExpress/AjaxControlToolkit/files/362742/HtmlEditor.zip)

**Comment by rclancy** [comment]
I haven't heard anything in 5 days since I uploaded the project...

**Comment by AlekseyMartynov** [comment]
We'll update this ticket as soon as we have any results.

**Comment by MikhailTymchukDX** [comment]
Yes, we have a problem with AjaxFileUpload and HtmlEditorExtender integration. While we can set AjaxFileUpload as a standalone control, we cannot change any property when it is a part of HtmlEditorExtender.

To work around this, you can set a handler path explicitly in the client-side code. For example, if your page URL is `http://localhost/HtmlEditor/WebForm1.aspx` (i.e. the application root is `http://localhost/HtmlEditor/`), then execute the following code:
```html
<body onload="load();">
...
    <script type="text/javascript">
        function load()
        {
            var extenderID = "<%= tbBody_HtmlEditorExtender.ClientID %>";
            $find(extenderID + "_ajaxFileUpload")._uploadUrl = "HtmlEditor/AjaxFileUploadHandler.axd";
        }
    </script>
</body>
```

I uploaded the modified version of your sample project with this code: [HtmlEditor2.zip](https://github.com/DevExpress/AjaxControlToolkit/files/373673/HtmlEditor2.zip)

**Comment by rclancy** [comment]
Thank you for the workaround and the modified project, although I couldn't get it to work properly in either in IISExpress or Localhost IIS. In both cases a file seems to upload correctly but is not displayed inside the editor.

What is the ETA for a fix in the AjaxControlToolkit? This is actually holding up development in more than one project of mine. I have to retain old versions of the toolkit for the apps to work as required.

**Comment by MikhailTymchukDX** [comment]
Hi @rclancy,
This fix is likely to be included in the post-16.1.1 release, because we need to make sure that it will not affect both HtmlEditorExtender and AjaxFileUpload behavior.
>I have to retain old versions of the toolkit for the apps to work as required.

Would you please specify the exact version of the toolkit where image uploading worked correctly?

In the meantime, I want to found out which step of the workaround I posted earlier does not work in your scenario:
1. Running some code after page loading.
2. Finding AjaxFileUpload that belongs to HtmlEditorExtender.
3. Setting the extender's variable to a specific value.

Would you please specify which step fails in your scenario?

**Comment by rclancy** [comment]
After trying several more recent versions several weeks ago, I found the latest I can use is v7.607 if I want to support multiple file uploads. However, with this version there is a bug: multiple file uploads only succeed when the page (using master pages) is located in a directory where there is no authentication required by a web.config at that location. I had to restructure my project to put all HtmlExtender > FileUploads in public directories.

None of the 3 steps you mention causes a problem. Without your fix the Upload button causes the file name to turn red and appends it with "(error)", as described in several posts not connected to the HtmlExtender (the fix for them being the location of the axd file). With your hotfix the filename turns green and appears successful, but when closing the dialog using the only available button "Cancel", no image is drawn in the textbox.

**Comment by MikhailTymchukDX** [comment]
> but when closing the dialog using the only available button "Cancel", no image is drawn in the textbox.

Do you see any errors in the browser's console (F12) after the upload?

**Comment by rclancy** [comment]
No

**Comment by MikhailTymchukDX** [comment]
@rclancy thank you, I did not mention that the image did not appear.

AjaxFileUpload has no code for image preview because it depends on the way the site stores images in every particular case. You can use the simple preview mode that is implemented in [Sample Site](http://ajaxcontroltoolkit.devexpress.com/AjaxFileUpload/AjaxFileUpload.aspx).

First, we [handle the `OnUploadComplete` event](https://github.com/DevExpress/AjaxControlToolkit/blob/master/AjaxControlToolkit.SampleSite/AjaxFileUpload/AjaxFileUpload.aspx.cs#L45-L49): save image contents to session and set the `PostedUrl` variable.
Client code [handles this event too](https://github.com/DevExpress/AjaxControlToolkit/blob/master/AjaxControlToolkit/Scripts/AjaxFileUpload.js#L1695-L1703) and uses `PostedUrl` to know what URL to fetch to show the image.
Finally, the server handles the image fetch request that contains the "preview" key and [writes image contents to the output](https://github.com/DevExpress/AjaxControlToolkit/blob/master/AjaxControlToolkit.SampleSite/AjaxFileUpload/AjaxFileUpload.aspx.cs#L14-L33).

I attached the modified version of your sample project that implements the image preview in that way, but you can change it according to your needs: [HtmlEditor3.zip](https://github.com/DevExpress/AjaxControlToolkit/files/393721/HtmlEditor3.zip)

**Comment by rclancy** [comment]
Sorry, I'd commented out the code behind ImageUploadComplete - my mistake. The image is now displayed.

Also, my comments previously about the authentication bug were concerned with FileUpload, not the HtmlExtender fileupload. I haven't tested this yet in later versions. I'm dealing with multiple projects, some use fileUpload, others HtmlExtender with upload.

I need to test your workaround with master pages. How would you go about this? The workaround is hard-coded for specific page elements and also uses the body tag, which is not available when using master pages.

When can we expect a release that fixes the axd location issue in both FileUpload and HtmlExtender upload?

**Comment by MikhailTymchukDX** [comment]
To apply this workaround to a master page, you can use [`Sys.Application.add_load` event](https://msdn.microsoft.com/en-us/library/bb383829.aspx).

Place the code, which does the following, to the master page:
1. Searches for HtmlEditorExtenders
2. Tries to get AjaxFileUpload behavior of the extender
3. Sets the `_uploadUrl` variable

Here is an example of such code:
```javascript
<script type="text/javascript">
    Sys.Application.add_load(function() {
        var components = Sys.Application.getComponents();

        for(var i = 0; i < components.length; i++) {
            var component = components[i];

            if(Sys.Extended.UI.HtmlEditorExtenderBehavior.isInstanceOfType(component))
                var ajaxFileUpload = $find(component.get_id() + "_ajaxFileUpload");
                if(ajaxFileUpload)
                    ajaxFileUpload._uploadUrl = "HtmlEditor/AjaxFileUploadHandler.axd";
        }
    });
</script>
```
I have attached an updated example of the sample project: [HtmlEditorFolder.zip](https://github.com/DevExpress/AjaxControlToolkit/files/395496/HtmlEditorFolder.zip)

Please take a note of some structural changes:
* Project folder is inside another folder
* Project URL is set to `http://localhost:60818/HtmlEditor`

Without the workaround applied, this will break HtmlEditorExtender. So, you can test this with IIS Express.

Another note: I'm using VS 2015, so IIS Express config has moved to the `.vs/config/applicationhost.config` file inside the project's folder. You' may have to tweak `<sites>/<site name="HtmlEditor" id="2">` element to set a correct physical path for the site.

>When can we expect a release that fixes the axd location issue in both FileUpload and HtmlExtender upload?

If things go smoothly this fix can be available in a few days.

**Comment by rclancy** [comment]
Mikhail, I really appreciate your help on this. As it will be only a matter of days and I have other pressing issues I would like to leave this for a week or two and hopefully the fix will be published in the meantime. Please include me on the release notification list. Thank you.

**Comment by MikhailTymchukDX** [comment]
@rclancy The build containing the fix can be downloaded from the [AppVeyor CI](https://ci.appveyor.com/project/dxrobot/ajaxcontroltoolkit/build/artifacts).

**Comment by rclancy** [comment]
Thank you. I usually install the toolkit using Nuget. Should I manually replace these dll's with the new build or wait for Nuget to offer the new build and upgrade normally?

**Comment by MikhailTymchukDX** [comment]
You can use any method depending on your needs.
I advise to replace DLLs manually and then wait for a release that includes this fix.

**Comment by rclancy** [comment]
I've updated the toolkit to 16.1.1 but I still can't get it to work. After selecting a file it is shown as "pending" but doesn't upload and I have to cancel the dialog. Is there an axd path setting that I need to set?

**Comment by MikhailTymchukDX** [comment]
@rclancy Yes, the InsertImage button provides now the new `AjaxFileUploadHandlerPath` property.
Set it to locate in the `AjaxFileUploadHandler.axd` handler path:
```xml
<ajaxToolkit:HtmlEditorExtender>
  <Toolbar>
    <ajaxToolkit:InsertImage AjaxFileUploadHandlerPath="MySite/AjaxFileUploadHandler.axd" />
  </Toolbar>
</ajaxToolkit:HtmlEditorExtender>
```

**Comment by rclancy** [comment]
It doesn't seem so with 16.1.1. This is the error message:
Type 'AjaxControlToolkit.InsertImage' does not have a public property named 'AjaxFileUploadHandlerPath'.

**Comment by MikhailTymchukDX** [comment]
This update was not included in the v16.1.1 release.
You have to [download build artifacts](https://ci.appveyor.com/project/dxrobot/ajaxcontroltoolkit/branch/master/artifacts) and [upgrade your project](https://github.com/DevExpress/AjaxControlToolkit/wiki/Updating-the-Project-from-CI-Builds).
